lib/keypair.rb

```diff
@@ -29,18 +42,23 @@ class Keypair < ActiveRecord::Base
 
   validates :_keypair, presence: true
   validates :jwk_kid, presence: true
+  validates :not_before, :expires_at, presence: true
+
+  validate :not_after_after_not_before
+  validate :expires_at_after_not_after
 
   after_initialize :set_keypair
+  after_initialize :set_validity
 
   # @!method valid
   #   @!scope class
-  #   The last 3 keypairs are considered valid and can be used to validate signatures and export public jwks.
-  #   It uses a subquery to make sure a +find_by+ actually searches only the valid 3 ones.
-  scope :valid, -> { where(id: unscoped.order(created_at: :desc).limit(3)) }
+  #   Non-expired keypairs are considered valid and can be used to validate signatures and export public jwks.
+  #   It uses a subquery to make sure a +find_by+ actually searches only the non-expired ones.
+  scope :valid, -> { where(id: unscoped.where('? < expires_at', Time.zone.now)) }
```

Comment:
I think the subquery here is not needed anymore. we needed this to make sure to only find the keypair if it was within the last three of the entire table and not of your query. 

Next to that I think it's good to use arel for these comparisons. [This is a nice blog about how to do that](https://jacopretorius.net/2016/09/the-mimimum-arel-every-rails-developer-should-know.html). This will especially be helpful when you are doing joins / eager loading etc since then rails knows the direct link to the table